### DISCUSSION TITLE
Evaluate using Link-Time Optimization (LTO) and Profile-Guided Optimization (PGO) for the project

### DISCUSSION THREAD

**Discussion by zamazan4ik** [acrlabs/simkube#131] [Feature Requests and Future Work]
Hi!

I want to discuss several compiler optimizations that could be helpful for the project. Maybe you even already tested them and have some benchmark numbers to share publicly ;)

Did you test Link-Time Optimization (LTO) for Simkube? If yes, do you have any numbers regarding performance and the resulting binary size? It would be nice to test since enabling LTO is not *too* time-consuming thing (usually, huh).

I suggest enabling LTO only for the Release builds so as not to sacrifice the developers' experience while working on the project since LTO consumes an additional amount of time to finish the compilation routine. If you think that a regular Release build should not be affected by such a change as well, then I suggest adding an additional `release-lto` profile where additionally to regular `release` optimizations LTO also will be added. Such a change simplifies life for maintainers and others interested in the project persons who want to build the most performant version of the application. Using ThinLTO also should help).

If we are talking about more advanced things, recently I started evaluating using Profile-Guided Optimization (PGO) for various applications and workloads (including different parsers and log solutions like Vector) - all the results are available at https://github.com/zamazan4ik/awesome-pgo .

Post-Link Optimization (PLO) is similar to the PGO technique but uses a slightly different approach. You can take a look at [LLVM BOLT](https://github.com/llvm/llvm-project/blob/main/bolt/README.md) for more information. However, I recommend starting to play with PLO only after applying PGO - PGO is a much more stable and time-proven technology compared to PLO tools.

Thank you.

**Comment by drmorr0** [comment]
Hi!  Thanks for your feedback!  I have not looked into LTO/PLO for this project at all.  Right now sadly there isn't even a difference between the "release" and "debug" builds, which is something I need to change, but haven't had a chance to set up yet (see #75).  Definitely for the release builds, using LTO or something similar would be desirable!

**Comment by drmorr0** [comment]
The "release" version of simkube now compiles with LTO turned on (see #206), it's quite time-consuming to build but I think will be an overall good thing for the project.  I haven't investigated any profiler-guided optimizations yet, I'm not sure that will add a ton of additional value.
